service/src/main/java/greencity/achievement/AchievementCalculation.java

```diff
@@ -0,0 +1,120 @@
+package greencity.achievement;
+
+import greencity.dto.achievement.AchievementVO;
+import greencity.dto.achievementcategory.AchievementCategoryVO;
+import greencity.dto.user.UserVO;
+import greencity.dto.useraction.UserActionVO;
+import greencity.entity.AchievementCategory;
+import greencity.entity.User;
+import greencity.entity.UserAchievement;
+import greencity.enums.AchievementType;
+import greencity.service.AchievementCategoryService;
+import greencity.service.AchievementService;
+import greencity.service.UserActionService;
+import greencity.service.UserService;
+import org.modelmapper.ModelMapper;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+
+import static greencity.enums.AchievementStatus.ACTIVE;
+import static greencity.enums.AchievementType.INCREMENT;
+
+@Component
+public class AchievementCalculation {
+    private UserService userService;
+    private UserActionService userActionService;
+    private AchievementService achievementService;
+    private AchievementCategoryService achievementCategoryService;
+    private final ModelMapper modelMapper;
+
+    /**
+     * {@inheritDoc} Constructor for {@link AchievementCalculation}
+     * 
+     * @param userService                {@link UserService}
+     * @param userActionService          {@link UserActionService}
+     * @param achievementService         {@link AchievementService}
+     * @param achievementCategoryService {@link AchievementCategoryService}
+     * @param modelMapper                {@link ModelMapper}
+     */
+    public AchievementCalculation(@Lazy UserService userService,
+        UserActionService userActionService,
+        @Lazy AchievementService achievementService,
+        AchievementCategoryService achievementCategoryService,
+        ModelMapper modelMapper) {
+        this.userService = userService;
+        this.userActionService = userActionService;
+        this.achievementService = achievementService;
+        this.achievementCategoryService = achievementCategoryService;
+        this.modelMapper = modelMapper;
+    }
```

Comment:
You can add final to instance variables and add annotation @RequiredArgsConstructor to remove constructor here